Service type changes made to FL Community Providers

gescobardev opened the pull request:
## Description
Resolves - https://github.com/department-of-veterans-affairs/va.gov-team/issues/26852
Also resolves - https://github.com/department-of-veterans-affairs/va.gov-team/issues/28734

Here is the [uxpin](https://preview.uxpin.com/602167f0ca1beccf4bdf706e04b3a5bd34c78e94#/pages/139907964/documentation/no-panels?mode=i) for reference

## Acceptance criteria

Per original issue: 

**Expected behavior for initial state**
- [ ] On input focus, with no text in the field, or fewer than two characters, a dropdown prompt appears: "Search for an available service"
- [ ] Screen reader announces prompt when it displays
- [ ] Input retains focus, prompt is not accessible via arrow keys

**Expected behavior if there is no service type match** 
- [ ] If no service type matches are found, a prompt is displayed: "We couldn't find that, please try another service"
- [ ] Screen reader announces prompt when it displays
- [ ] Input retains focus, prompt is not accessible via arrow keys

**Expected behavior if search is attempted with no valid service selection** 
- [ ] If search is attempted with no valid service selection, a prompt is displayed: "Please search for an available service." and no search is executed
- [ ] Current screen reader and focus management for this scenario is maintained 

## Definition of done
- [ ] Events are logged appropriately
- [ ] Documentation has been updated, if applicable
- [ ] A link has been provided to the originating GitHub issue (or connected to it via ZenHub)
- [ ] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs

Commit:
service type changes made to Community Providers

Commit:
service type changes made to Community Providers

Commit:
skip failing tests

Commit:
skip failing tests

mpelzsherman commented on src/applications/facility-locator/components/ServiceTypeAhead.jsx:
```diff
@@ -103,38 +143,50 @@ class ServiceTypeAhead extends Component {
+                        </div>
+                      ))}
+                  </div>
+                )}
```
What do you think about moving the block that renders this <div> into a function, to improve readability/organization?

gescobardev replied:
good suggestion

mpelzsherman commented on src/applications/facility-locator/components/ServiceTypeAhead.jsx:
```diff
@@ -62,6 +65,43 @@ class ServiceTypeAhead extends Component {
+
+    if (
+      (isFocused && !inputValue && !showError) ||
+      (inputValue && inputValue.length < 2)
```
Suggestion: define a local constant, e.g. `MIN_SEARCH_CHARS`, and use it here and also on line 167 in place of the hard-coded number "2". Makes the code more readable and makes it easier to change later.

gescobardev replied:
sounds good to me

mpelzsherman reviewed (changes requested):
At minimum, we should fix the existing e2e tests, and add an e2e test to src/applications/facility-locator/tests/e2e/errorMessages.cypress.spec.js where we try to click “search” with an invalid service type.

Additional unit tests can be added in another PR.

mpelzsherman commented on src/applications/facility-locator/components/ServiceTypeAhead.jsx:
```diff
@@ -4,6 +4,7 @@ import { connect } from 'react-redux';
 import Downshift from 'downshift';
 import { getProviderSpecialties } from '../actions';
 import classNames from 'classnames';
+// import environment from 'platform/utilities/environment';
```
Delete pls

mpelzsherman committed:
Fixed cypress tests

Formatted JSON file

mpelzsherman committed:
Fixed cypress tests

Formatted JSON file

mpelzsherman commented on src/applications/facility-locator/components/ServiceTypeAhead.jsx:
```diff
@@ -62,6 +65,43 @@ class ServiceTypeAhead extends Component {
+    if (
+      inputValue &&
+      inputValue.length >= 2 &&
+      !services.filter(specialty => this.shouldShow(inputValue, specialty))
```
We might want to refactor this into a function:

```js
matchingServices = inputValue =>
      this.state.services.filter(specialty => this.shouldShow(inputValue, specialty))
```

.. and use it on line 169 as well.

gescobardev replied:
I can do this

mpelzsherman commented on src/applications/facility-locator/components/ServiceTypeAhead.jsx:
```diff
@@ -62,6 +65,43 @@ class ServiceTypeAhead extends Component {
+          <div className="vads-u-margin--1p5">
+            {"We couldn't find that, please try another service"}
+          </div>
+        </div>
```
Consider refactoring this block into a function, e.g.

```js
  renderMessageDiv = message => (
       <div className="dropdown" role="alert">
          <div className="vads-u-margin--1p5">
            {message}
          </div>
        </div>
  );
```

Use in renderSearchForAvailableServicePrompt.

Could even be its own mini-component!

gescobardev replied:
I like the mini component idea!

mpelzsherman commented:
Added [ticket](https://github.com/department-of-veterans-affairs/va.gov-team/issues/28688) for adding unit test coverage.

Commit:
made suggested changes

Commit:
made suggested changes

Commit:
adjusted max-height of dropdown

Commit:
adjusted max-height of dropdown

Commit:
fixed issues with search for available service prompt

Commit:
fixed issues with search for available service prompt

mpelzsherman committed:
Remove warnings about controlled inputs, and disable input while fetching svcs

mpelzsherman committed:
Remove warnings about controlled inputs, and disable input while fetching svcs

Commit:
simplified renderSearchForAvailableServicePrompt if statement

Commit:
simplified renderSearchForAvailableServicePrompt if statement

mpelzsherman commented:
I would suggest adding a few Cypress tests:

1. Confirm that the "Search for an available service" message is visible when the Service Type field is focussed
2. Confirm that the "We couldn't find that" message is visible after typing 3 chars with no results
3. Confirm that the form can't be submitted without selecting a service type (error message is displayed)
4. Confirm you can select a service type and click "search"; search results are displayed and service type is shown in the results header

Commit:
added cypress tests

Commit:
added cypress tests

Commit:
cleaned code

Commit:
cleaned code

gescobardev commented:
@mpelzsherman cypress tests added

mpelzsherman committed:
Remove unnecessary hard-coded wait().

mpelzsherman committed:
Remove unnecessary hard-coded wait().

mpelzsherman commented on src/applications/facility-locator/tests/e2e/providerSearch.cypress.spec.js:
```diff
@@ -11,6 +11,41 @@ describe('Provider search', () => {
 
+  it('renders "Search for available service" prompt', () => {
+    cy.visit('/find-locations');
+    cy.injectAxe();
```
I think this can be removed since we aren't running an axe check.

gescobardev replied:
removed

mpelzsherman commented on src/applications/facility-locator/tests/e2e/providerSearch.cypress.spec.js:
```diff
@@ -11,6 +11,41 @@ describe('Provider search', () => {
+
+  it("renders `We couldn't find that, please try another service ` prompt", () => {
+    cy.visit('/find-locations');
+    cy.injectAxe();
```
I think this can be removed since we aren't running an axe check.

gescobardev replied:
removed

Commit:
remove injectAxe() command

Commit:
remove injectAxe() command

Commit:
Merge branch 'fl-service-type-changes-26852' of https://github.com/department-of-veterans-affairs/vets-website into fl-service-type-changes-26852

Commit:
removed onBlur prop

Commit:
removed onBlur prop

Commit:
fixed keyboard nav service type select to search issue

Commit:
fixed keyboard nav service type select to search issue

Commit:
removed handleFacilityTypeOnBlur()

Commit:
removed handleFacilityTypeOnBlur()

mpelzsherman commented on src/applications/facility-locator/components/ServiceTypeAhead.jsx:
```diff
@@ -103,38 +171,34 @@ class ServiceTypeAhead extends Component {
+                  this.setState({ isFocused: false });
+                  // this.props.onBlur(e);
+                }}
+                // onFocus={() => this.setState({ isFocused: true })}
```
Please delete comments

gescobardev replied:
removed

mpelzsherman commented on src/applications/facility-locator/tests/e2e/errorMessages.cypress.spec.js:
```diff
@@ -43,13 +43,27 @@ describe('Facility search error messages', () => {
+    // Wait for services to be saved to state and input field to not be disabled
+    // eslint-disable-next-line cypress/no-unnecessary-waiting
+    cy.wait(600);
     cy.get('#service-type-ahead-input').focus();
```
Try adding `.should('not.be.disabled')` - hopefully we can remove the hard-coded `wait` and the eslint-disable.

gescobardev replied:
This worked, thanks!

Commit:
removed commmented code, added and updated e2e tests

Commit:
removed commmented code, added and updated e2e tests

Commit:
fixed issue onInputValueChange

Commit:
fixed issue onInputValueChange

mpelzsherman committed:
Add cypress test for deleting service after search

Also removed handleServiceTypeBlur from SearchControls.jsx
since this is being handled by ServiceTypeAhead now.

mpelzsherman committed:
Add a delay before rendering the "We couldn't find that" error message

Rleahy22 commented on src/applications/facility-locator/components/MessagePromptDiv.jsx:
```diff
@@ -0,0 +1,28 @@
+
+MessagePromptDiv.propTypes = {
+  id: PropTypes.string,
+  message: PropTypes.string,
```
Should `waitBeforeShow` get added to `propTypes`?

gescobardev replied:
Yes, I can add it. 
Thanks @Rleahy22 !

Rleahy22 commented on src/applications/facility-locator/tests/e2e/errorMessages.cypress.spec.js:
```diff
@@ -30,8 +30,10 @@ describe('Facility search error messages', () => {
+    cy.get('#street-city-state-zip').type('Austin, TX');
     cy.get('#facility-type-dropdown').focus();
-    cy.get('#street-city-state-zip').focus();
+    // cy.get('#street-city-state-zip').focus();
```
Can this be deleted?

gescobardev replied:
Line has been removed

Rleahy22 reviewed (approved):
Just small comments, but LGTM!

Commit:
Merge branch 'fl-service-type-changes-26852' of https://github.com/department-of-veterans-affairs/vets-website into fl-service-type-changes-26852

Commit:
Merge branch 'fl-service-type-changes-26852' of https://github.com/department-of-veterans-affairs/vets-website into fl-service-type-changes-26852

Commit:
removed commented code and added propTypes for waitBeforeShow

Commit:
accessibility changes made with error message and could-not-find-service prompt
